### ISSUE TITLE
Update New Contributor Documentation

### ISSUE THREAD

**Issue by josephj11** [autokey/autokey#718]
### Has this issue already been reported?

- [X] I have searched through the existing issues.

### Is this a question rather than an issue?

- [X] This is not a question.

### What type of issue is this?

Documentation

### Can you briefly describe the issue?

Contributor pages need updating now that 0.96.0 has been released.

Remove references to 0.95.10 from https://github.com/autokey/autokey/blob/develop/CONTRIBUTORS.rst#testing

https://github.com/autokey/autokey/wiki/Contributing-code needs to be looked at as well.

One of the developers should probably do this so no inaccuracies are introduced.

**Comment by Elliria** [comment]
The **CONTRIBUTORS.rst** file also contains this section:

> Please make sure tests pass before you submit PRs. To ensure this happens automatically, I recommend adding the following lines to the file .git/hooks/pre-push:

One would expect that to be followed by some lines of code, but there's no code. That should be updated as well.

Update: I did fix this part already.

**Comment by Elliria** [comment]
Also related to this is discussion #805.

**Comment by Elliria** [comment]
Please rename this to: **Update CONTRIBUTORS.rst**

**Comment by Elliria** [comment]
I'm adding your suggestion from elsewhere here, @josephj11: 

Change **we** to **I** by replacing this:	

> Just open a PR for discussion and I’ll try to answer any questions that arise. I suggest writing new features on top of develop

With this:

> Just open a PR against the **develop** branch for discussion and we’ll try to answer any questions that arise. We prefer to leave our **master** branch relatively untouched between releases as a reference for current issues and user support.

**Comment by Elliria** [comment]
I also don't fully understand everything in the CONTRIBUTORS.rst file yet, so some of it will have to wait until I've gotten more of an education.

Meanwhile, I'll try to follow the suggestion [here](https://github.com/autokey/autokey/discussions/805) for doing diffs to maybe figure out the answer myself or wait for a reply to that from someone who knows the answer before deciding whether or not to remove this section from the document:

> The current (0.95.x) master tests are deprecated and won’t work. They are still on Python 2 and nobody cared to update them throughout the years, not even the original developer. They test things long gone in the project code… The old tests are scheduled to be replaced by a new suite in the develop branch, which will be merged as 0.96.0.

> The current situation is a bit unfortunate: The new suite lives in the develop branch, which accumulates new features, but which can't be backported to master without introducing merge conflicts later on. So until develop gets merged in, you’d have to switch to develop to run the tests.

**Comment by josephj11** [comment]
No rush. We want to get it right the first try whenever possible.

Simplistically looking at it: Develop works. So make a copy of main and slap the testing, etc, into it and see what happens. At best, it just works - maybe with a few links and labels updated. At worst, it breaks badly, but gives us some clues as to what needs to be fixed. This assumes that you can figure out the "slap" part without too much trouble.

**Comment by Elliria** [comment]
I think when you have a point in development where you're happy with **develop**, that's when you create **beta** and copy what you like from **develop** into it (all of **develop** if you're not unhappy with any of it). Then **beta** goes through whatever testing it can get before it's decided that, yep, it's good before overwriting **master** with it and calling it the new release.

So far, my research shows that the slapping would be done by using the "ours" merge strategy to overwrite **master** with **beta**, but this would definitely need to be tested out in a dummy repository first:

```bash
# Get the beta branch:
git checkout beta
# Instruct git to discard all changes from another branch in the next merge
# with master and only consider changes made from this point forward:
git merge -s ours master
# Get the master branch:
git checkout master
# Merge beta with master:
git merge beta
```

**Comment by josephj11** [comment]
I don't know what the ours options are. However, my concept (possibly because I'm not that into git) is that each release is self contained. Develop has everything it needs. When ready for release, it should be copied to  a new empty branch named beta. Similarly, when beta is ready, it completely replaces main after main has been saved as a prior version (which should have happened when main was first released and only needs to be done again if we have touched master after release). There are no merges.

If merges are intended to catch post release changes to main, that's something I want to avoid. There could be some sort of compare or merge purely as a test to verify that everything that has actually changed is exactly what we think it is, nothing more, nothing less.

Any merges should be from main to develop to catch modifications made directly to main. This should be done as soon as possible after any post release change to main, long before beta exists.

**Comment by Elliria** [comment]
I'm good with doing things however you'd like with the first part of that, especially since you're doing the actual merges. So, do your thing and I'll just work with whatever is there.

I'm not sure if I understand this part, though:

> Any merges should be from main to develop to catch modifications made directly to main. This should be done as soon as possible after any post release change to main, long before beta exists.

I thought the whole point of **develop** was that that's where we try things. If we like them and they're needed in **master**, then we do them in **master** as well. Have I got that backwards?

**Comment by josephj11** [comment]
This is exactly why I was saying not to do PR's on main. It makes things confusing.

If, for instance, you do a PR of the code of conduct to main. That one file will be ahead of develop. Mergeing backwards from main to develop would fix that, but you'd probably have to review the merge line by line to make sure that's all that gets merged. If done wrong, old code in main would overwrite new code in develop.

It would be cool (so someone has probably already done it) if you could make the merge conditional such that only changes/differences newer than some date (such as the previous release date) are considered for merging. This could still overwrite newer code, but there would be a whole lot less changes to review.

You could do a diff of current main against released main to find all the candidate code, but it might take a bit of advanced git fu to then "pipe" that result into the backward merge test run  and then, once sanitized, into the actual merge into develop.

It can't be rocket science because, AFAIK, all these changes are just the automated application of diff files. It could be quite complicated in terms of details, but conceptually, it's very simple.

**Comment by Elliria** [comment]
I see what you mean. This is a horrendous mess. If you look at our branches, each branch is X number of commits ahead of **and** X number of commits behind **master** and none of them are identical [sigh]. I have no idea how one would sort something like that out. If it was done one commit at a time, it would take forever.
***********************************************************************************************
I'm now a **firm** believer of your original philosophy of "hands off of **master**". Consider me a complete convert. I think we should make that a hard-and-fast rule in the future after the very next release. So, in future, our development would look like this:
* Whatever **master** looks like at the moment of the latest release should be what it stays like from then on until the next release.
* Only **develop** and any other feature branches should change.
* If **master** needs to change because something significant is wrong with it that's been fixed elsewhere, then it should be time for a point-release, at which point **master** would again freeze in that new state until the next release of any kind.
 
That sounds so much neater and simpler than what we're currently dealing with. No branches would ever be behind **master** and all we would ever need to know is in what way any of the other branches are ahead of **master**.

**Comment by Elliria** [comment]
Please rename this issue to: **Update CONTRIBUTORS.rst**

**Comment by josephj11** [comment]
Sorting it out: When it gets like that you don't try to fix it unless paid highly to do so. Any branches we keep are history. If you're working on  piece of code and  it doesn't look right, etc., you can diff it with the historic versions to see if there's more to the story.

**Comment by Elliria** [comment]
Yeah, it seems that this is a challenging, yet interesting, place to climb around in.
